great_expectations/experimental/datasources/sql_datasource.py

```diff
@@ -207,73 +161,84 @@ def batch_request_options_template(
             return template
         return {p: None for p in self.column_splitter.param_names}
 
-    # This asset type will support a variety of splitters
-    def add_year_and_month_splitter(
+    def _add_splitter(self, method_name: str, column_name: str):
+        self.column_splitter = ColumnSplitter(
+            column_name=column_name,
+            method_name=method_name,
+        )
+        self.test_column_splitter_connection()
+        return self
+
+    def add_splitter_year(
+        self,
+        column_name: str,
+    ) -> _SQLAsset:
+        """Associates a year splitter with this sql data asset.
+        Args:
+            column_name: A column name of the date column where year and month will be parsed out.
+        Returns:
+            This sql asset so we can use this method fluently.
+        """
+        return self._add_splitter(method_name="split_on_year", column_name=column_name)
+
+    def add_splitter_year_and_month(
+        self,
+        column_name: str,
+    ) -> _SQLAsset:
+        """Associates a year, month splitter with this sql asset.
+        Args:
+            column_name: A column name of the date column where year and month will be parsed out.
+        Returns:
+            This sql asset so we can use this method fluently.
+        """
+        return self._add_splitter(
+            method_name="split_on_year_and_month", column_name=column_name
+        )
+
+    def add_splitter_year_and_month_and_day(
         self,
         column_name: str,
     ) -> _SQLAsset:
-        """Associates a year month splitter with this _SQLAsset
+        """Associates a year, month, day splitter with this sql asset.
         Args:
             column_name: A column name of the date column where year and month will be parsed out.
         Returns:
-            This _SQLAsset so we can use this method fluently.
+            This sql asset so we can use this method fluently.
         """
-        raise NotImplementedError(
-            """One needs to implement "add_year_and_month_splitter" on a _SQLAsset subclass."""
+        return self._add_splitter(
+            method_name="split_on_year_and_month_and_day", column_name=column_name
         )
 
-    def _fully_specified_batch_requests(self, batch_request) -> List[BatchRequest]:
+    def test_connection(self) -> None:
+        pass
+
+    def test_column_splitter_connection(self) -> None:
+        pass
```

Comment:
I would think this shouldn't be necessary if we define `test_connection` on the splitter itself. The existing pattern is something like:
```
def test_connection(self) -> None:
    ...
    if self.column_splitter:
        self.column_splitter.test_connection()
```

Reply:
I almost removed ColumnSplitter.test_connection all together to put that logic in the TableAsset. I feel that `test_connection` on the column splitter is in the wrong level on abstraction since it doesn't know about the database or it's connection. The same column splitter should also work on arbitrary/all asset types. If an implementer is adding a new `_SQLDataAsset` subclass, they should be able to subclass `_SQLDataAsset` look at the methods that need to be implemented and be done. They shouldn't have to peer at the rest of our code and it's implementation. Currently, they have to subclass _`SQLDataAsset`, realize `ColumnSplitter.test_connection` is called and takes a parameter of the `DataAsset` type. They then have to modify `ColumnSplitter` and overload `test_connect` to be `test_connection(self, my_new_asset)` likely by adding a new `@overload` decorated empty method and updating the actual implementation. By making `test_column_splitter_connection` explicit on the `DataAsset`, the subclasser will know what they have to implement. I didn't remove `test_connection` on the ColumnSplitter since we are only testing it on table assets currently but I wanted to be explicit about what a DataAsset subclasser would have to implement.

Reply:
This makes sense to me, but it means we probably want to make a similar change in `Datasource.test_connection` for `assets`.